server/configuration.go

```diff
@@ -22,9 +23,43 @@ import (
 // If you add non-reference types to your configuration struct, be sure to rewrite Clone as a deep
 // copy appropriate for your types.
 type configuration struct {
-	OAuth2Authority    string
-	OAuth2ClientID     string
-	OAuth2ClientSecret string
+	OAuth2Authority    string `json:"oauth2authority"`
+	OAuth2ClientID     string `json:"oauth2clientid"`
+	OAuth2ClientSecret string `json:"oauth2clientsecret"`
+	EncryptionKey      string `json:"encryptionkey"`
+}
+
+func (c *configuration) ToMap() (map[string]interface{}, error) {
+	var out map[string]interface{}
+	data, err := json.Marshal(c)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(data, &out)
+	if err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+func (c *configuration) setDefaults() (bool, error) {
+	if c.OAuth2ClientID != "" {
+		return false, nil
+	}
+
+	changed := false
+	if c.EncryptionKey == "" {
+		secret, err := generateSecret()
+		if err != nil {
+			return false, err
+		}
+
+		c.EncryptionKey = secret
+		changed = true
+	}
```

Comment:
This makes it so all user keys are invalidated upon upgrading the plugin. Is this the route we want to take?

Reply:
cc @aaronrothschild @DHaussermann for input on this blocking question

Reply:
How can an admin force user's to get a new token if they want to?  

I agree, would prefer that the tokens are refreshed and then encrypted if possible.

Reply:
@mickmister @levb Can we keep the current encryption key, and if the admin ever wants to update it, they can do it via the system console?  Would this avoid all user's needing to re-authenticate after an update?

cc @DHaussermann

Reply:
@aaronrothschild On plugin upgrade, there is no "current encryption key". At the moment, upgrading the plugin automatically creates a new encryption key since it is not defined.

I'm suggesting we shouldn't define it automatically on upgrade, and instead have the admin explicitly generate the token in the system console, which would only then cause all users' tokens to be reset.

Reply:
Double-checked the logic. The setting of the default encryption key should only happen on installations where the Client ID had not been configured, meaning they're unused.

Existing installations will not be auto-encrypted/invalidated. But if the key is regenerated, all user records will be wiped out.